File: sdks/typescript/src/types/vre.ts
/**
 * VRE (Visual Response Engine) Types
 *
 * Types for rendering visual templates that can be sent to messaging channels.
 */

/**
 * Output format for rendered images
 */
export type VREOutputFormat = 'png' | 'webp' | 'jpeg';

/**
 * Channel type for VRE rendering (affects image optimization)
 */
export type VREChannelType = 'whatsapp' | 'telegram' | 'web' | 'email';

/**
 * Available template types
 */
export type VRETemplateType =
  | 'menu_opcoes'
  | 'card_produto'
  | 'status_pedido'
  | 'lista_produtos'
  | 'confirmacao'
  | 'cobranca_pix';

/**
 * Order status for status_pedido template
 */
export type OrderStatus =
  | 'recebido'
  | 'separacao'
  | 'faturado'
  | 'transporte'
  | 'entregue';

/**
 * Stock status for products
 */
export type StockStatus = 'disponivel' | 'baixo' | 'indisponivel';

// ============================================
// Render Request/Response
// ============================================

/**
 * Request to render a VRE template
 */
export interface VRERenderRequest {
  /**
   * Tenant ID
   */
  tenantId: string;

  /**
   * Template ID to render
   */
  templateId: VRETemplateType;

  /**
   * Template data (varies by template type)
   */
  data: Record<string, unknown>;

  /**
   * Target channel for optimized rendering
   * @default 'whatsapp'
   */
  channel?: VREChannelType;

  /**
   * Output image format
   * @default 'webp'
   */
  format?: VREOutputFormat;

  /**
   * Override default width
   */
  width?: number;

  /**
   * Quality (0-100, only for webp/jpeg)
   */
  quality?: number;

  /**
   * Scale factor (1.0-2.0)
   */
  scale?: number;
}

/**
 * Response from rendering a template
 */
export interface VRERenderResponse {
  /**
   * Base64-encoded image data
   */
  imageBase64: string;

  /**
   * Auto-generated caption for the image
   */
  caption: string;

  /**
   * Image width in pixels
   */
  width: number;

  /**
   * Image height in pixels
   */
  height: number;

  /**
   * Output format used
   */
  format: VREOutputFormat;

  /**
   * Render time in milliseconds
   */
  renderTimeMs: number;

  /**
   * Image size in bytes
   */
  sizeBytes?: number;

  /**
   * Whether the result was served from cache
   */
  cacheHit?: boolean;
}

// ============================================
// Render and Send
// ============================================

/**
 * Request to render a template and send it directly to a conversation
 */
export interface VRERenderAndSendRequest {
  /**
   * Conversation ID to send the rendered image to
   */
  conversationId: string;

  /**
   * Template ID to render
   */
  templateId: VRETemplateType;

  /**
   * Template data
   */
  data: Record<string, unknown>;

  /**
   * Optional custom caption (auto-generated if not provided)
   */
  caption?: string;

  /**
   * Optional text message to send after the image
   */
  followUpText?: string;
}

/**
 * Response from render and send
 */
export interface VRERenderAndSendResponse {
  /**
   * ID of the sent message
   */
  messageId: string;

  /**
   * URL of the uploaded image
   */
  imageUrl: string;

  /**
   * Caption used
   */
  caption: string;
}

// ============================================
// Templates
// ============================================

/**
 * VRE template definition
 */
export interface VRETemplate {
  /**
   * Template ID
   */
  id: VRETemplateType;

  /**
   * Human-readable name
   */
  name: string;

  /**
   * Template description
   */
  description: string;

  /**
   * JSON schema for template data
   */
  schema: Record<string, unknown>;
}

/**
 * Response from listing templates
 */
export interface VREListTemplatesResponse {
  templates: VRETemplate[];
}

/**
 * Request to preview a template
 */
export interface VREPreviewRequest {
  /**
   * Template ID to preview
   */
  templateId: VRETemplateType;

  /**
   * Optional custom data (uses default sample data if not provided)
   */
  data?: Record<string, unknown>;
}

/**
 * Response from template preview
 */
export interface VREPreviewResponse {
  /**
   * Base64-encoded image data
   */
  imageBase64: string;

  /**
   * Image width
   */
  width: number;

  /**
   * Image height
   */
  height: number;
}

// ============================================
// Template Data Types
// ============================================

/**
 * Menu option for menu_opcoes template
 */
export interface MenuOpcaoData {
  /**
   * Option label text
   */
  label: string;

  /**
   * Optional description
   */
  descricao?: string;

  /**
   * Icon name (pedido, catalogo, entrega, financeiro, atendente, etc.)
   */
  icone?: string;
}

/**
 * Data for menu_opcoes template
 */
export interface MenuOpcoesData {
  /**
   * Menu title
   */
  titulo: string;

  /**
   * Optional subtitle
   */
  subtitulo?: string;

  /**
   * Menu options (max 8)
   */
  opcoes: MenuOpcaoData[];

  /**
   * Optional message to show before options
   */
  mensagemAntes?: string;
}

/**
 * Data for card_produto template
 */
export interface CardProdutoData {
  /**
   * Product name
   */
  nome: string;

  /**
   * Product SKU
   */
  sku?: string;

  /**
   * Product price
   */
  preco: number;

  /**
   * Unit (kg, un, cx, fd, pc)
   */
  unidade: string;

  /**
   * Stock quantity
   */
  estoque?: number;

  /**
   * Product image URL
   */
  imagemUrl?: string;

  /**
   * Highlight badge (promocao, novo, mais vendido)
   */
  destaque?: string;

  /**
   * Additional message
   */
  mensagem?: string;
}

/**
 * Timeline step for status_pedido template
 */
export interface StatusPedidoStep {
  /**
   * Step status (done, active, wait)
   */
  status: 'done' | 'active' | 'wait';

  /**
   * Step icon
   */
  icon: string;

  /**
   * Step label
   */
  label: string;
}

/**
 * Data for status_pedido template
 */
export interface StatusPedidoData {
  /**
   * Order number
   */
  numeroPedido: string;

  /**
   * Current order status
   */
  statusAtual: OrderStatus;

  /**
   * Items summary
   */
  itensResumo?: string;

  /**
   * Total order value
   */
  valorTotal?: number;

  /**
   * Delivery estimate
   */
  previsaoEntrega?: string;

  /**
   * Driver name
   */
  motorista?: string;

  /**
   * Additional message
   */
  mensagem?: string;
}

/**
 * Product item for lista_produtos template
 */
export interface ListaProdutoItem {
  /**
   * Product name
   */
  nome: string;

  /**
   * Product price
   */
  preco: number;

  /**
   * Unit
   */
  unidade?: string;

  /**
   * Stock status
   */
  estoqueStatus?: StockStatus;

  /**
   * Product SKU
   */
  sku?: string;

  /**
   * Emoji for the product
   */
  emoji?: string;
}

/**
 * Data for lista_produtos template
 */
export interface ListaProdutosData {
  /**
   * List title
   */
  titulo: string;

  /**
   * Products (max 6)
   */
  produtos: ListaProdutoItem[];

  /**
   * Additional message
   */
  mensagem?: string;
}

/**
 * Confirmation item for confirmacao template
 */
export interface ConfirmacaoItem {
  /**
   * Item name
   */
  nome: string;

  /**
   * Quantity description
   */
  quantidade?: string;

  /**
   * Item price
   */
  preco: number;

  /**
   * Emoji
   */
  emoji?: string;
}

/**
 * Data for confirmacao template
 */
export interface ConfirmacaoData {
  /**
   * Confirmation title
   */
  titulo?: string;

  /**
   * Subtitle
   */
  subtitulo?: string;

  /**
   * Items
   */
  itens: ConfirmacaoItem[];

  /**
   * Total value
   */
  valorTotal: number;

  /**
   * Delivery estimate
   */
  previsaoEntrega?: string;

  /**
   * Additional message
   */
  mensagem?: string;
}

/**
 * Data for cobranca_pix template
 */
export interface CobrancaPixData {
  /**
   * Payment amount
   */
  valor: number;

  /**
   * Order number
   */
  numeroPedido?: string;

  /**
   * PIX EMV/BRCode payload
   */
  pixPayload: string;

  /**
   * Expiration time (e.g., "30 minutos")
   */
  expiracao?: string;

  /**
   * Additional message
   */
  mensagem?: string;
}

File: sdks/go/types/vre.go
package types

// VREOutputFormat for rendered images
type VREOutputFormat string

const (
	VREOutputFormatPNG  VREOutputFormat = "png"
	VREOutputFormatWebP VREOutputFormat = "webp"
	VREOutputFormatJPEG VREOutputFormat = "jpeg"
)

// VREChannelType for VRE rendering
type VREChannelType string

const (
	VREChannelWhatsApp VREChannelType = "whatsapp"
	VREChannelTelegram VREChannelType = "telegram"
	VREChannelWeb      VREChannelType = "web"
	VREChannelEmail    VREChannelType = "email"
)

// VRETemplateType available templates
type VRETemplateType string

const (
	VRETemplateMenuOpcoes    VRETemplateType = "menu_opcoes"
	VRETemplateCardProduto   VRETemplateType = "card_produto"
	VRETemplateStatusPedido  VRETemplateType = "status_pedido"
	VRETemplateListaProdutos VRETemplateType = "lista_produtos"
	VRETemplateConfirmacao   VRETemplateType = "confirmacao"
	VRETemplateCobrancaPix   VRETemplateType = "cobranca_pix"
)

// OrderStatus for status_pedido template
type OrderStatus string

const (
	OrderStatusRecebido   OrderStatus = "recebido"
	OrderStatusSeparacao  OrderStatus = "separacao"
	OrderStatusFaturado   OrderStatus = "faturado"
	OrderStatusTransporte OrderStatus = "transporte"
	OrderStatusEntregue   OrderStatus = "entregue"
)

// StockStatus for products
type StockStatus string

const (
	StockStatusDisponivel   StockStatus = "disponivel"
	StockStatusBaixo        StockStatus = "baixo"
	StockStatusIndisponivel StockStatus = "indisponivel"
)

// VRERenderRequest for rendering a template
type VRERenderRequest struct {
	TenantID   string                 `json:"tenant_id"`
	TemplateID VRETemplateType        `json:"template_id"`
	Data       map[string]interface{} `json:"data"`
	Channel    VREChannelType         `json:"channel,omitempty"`
	Format     VREOutputFormat        `json:"format,omitempty"`
	Width      int                    `json:"width,omitempty"`
	Quality    int                    `json:"quality,omitempty"`
	Scale      float64                `json:"scale,omitempty"`
}

// VRERenderResponse from rendering a template
type VRERenderResponse struct {
	ImageBase64  string          `json:"image_base64"`
	Caption      string          `json:"caption"`
	Width        int             `json:"width"`
	Height       int             `json:"height"`
	Format       VREOutputFormat `json:"format"`
	RenderTimeMs int             `json:"render_time_ms"`
	SizeBytes    int64           `json:"size_bytes,omitempty"`
	CacheHit     bool            `json:"cache_hit,omitempty"`
}

// VRERenderAndSendRequest for render and send
type VRERenderAndSendRequest struct {
	ConversationID string                 `json:"conversation_id"`
	TemplateID     VRETemplateType        `json:"template_id"`
	Data           map[string]interface{} `json:"data"`
	Caption        string                 `json:"caption,omitempty"`
	FollowUpText   string                 `json:"follow_up_text,omitempty"`
}

// VRERenderAndSendResponse from render and send
type VRERenderAndSendResponse struct {
	MessageID string `json:"message_id"`
	ImageURL  string `json:"image_url"`
	Caption   string `json:"caption"`
}

// VRETemplate definition
type VRETemplate struct {
	ID          VRETemplateType        `json:"id"`
	Name        string                 `json:"name"`
	Description string                 `json:"description"`
	Schema      map[string]interface{} `json:"schema"`
}

// VREListTemplatesResponse from listing templates
type VREListTemplatesResponse struct {
	Templates []VRETemplate `json:"templates"`
}

// VREPreviewRequest for previewing a template
type VREPreviewRequest struct {
	TemplateID VRETemplateType        `json:"template_id"`
	Data       map[string]interface{} `json:"data,omitempty"`
}

// VREPreviewResponse from preview
type VREPreviewResponse struct {
	ImageBase64 string `json:"image_base64"`
	Width       int    `json:"width"`
	Height      int    `json:"height"`
}

// MenuOpcaoData for menu_opcoes template
type MenuOpcaoData struct {
	Label     string `json:"label"`
	Descricao string `json:"descricao,omitempty"`
	Icone     string `json:"icone,omitempty"`
}

// MenuOpcoesData for menu_opcoes template
type MenuOpcoesData struct {
	Titulo        string          `json:"titulo"`
	Subtitulo     string          `json:"subtitulo,omitempty"`
	Opcoes        []MenuOpcaoData `json:"opcoes"`
	MensagemAntes string          `json:"mensagem_antes,omitempty"`
}

// CardProdutoData for card_produto template
type CardProdutoData struct {
	Nome      string  `json:"nome"`
	SKU       string  `json:"sku,omitempty"`
	Preco     float64 `json:"preco"`
	Unidade   string  `json:"unidade"`
	Estoque   int     `json:"estoque,omitempty"`
	ImagemURL string  `json:"imagem_url,omitempty"`
	Destaque  string  `json:"destaque,omitempty"`
	Mensagem  string  `json:"mensagem,omitempty"`
}

// StatusPedidoStep for status_pedido template
type StatusPedidoStep struct {
	Status string `json:"status"`
	Icon   string `json:"icon"`
	Label  string `json:"label"`
}

// StatusPedidoData for status_pedido template
type StatusPedidoData struct {
	NumeroPedido    string      `json:"numero_pedido"`
	StatusAtual     OrderStatus `json:"status_atual"`
	ItensResumo     string      `json:"itens_resumo,omitempty"`
	ValorTotal      float64     `json:"valor_total,omitempty"`
	PrevisaoEntrega string      `json:"previsao_entrega,omitempty"`
	Motorista       string      `json:"motorista,omitempty"`
	Mensagem        string      `json:"mensagem,omitempty"`
}

// ListaProdutoItem for lista_produtos template
type ListaProdutoItem struct {
	Nome          string      `json:"nome"`
	Preco         float64     `json:"preco"`
	Unidade       string      `json:"unidade,omitempty"`
	EstoqueStatus StockStatus `json:"estoque_status,omitempty"`
	SKU           string      `json:"sku,omitempty"`
	Emoji         string      `json:"emoji,omitempty"`
}

// ListaProdutosData for lista_produtos template
type ListaProdutosData struct {
	Titulo   string             `json:"titulo"`
	Produtos []ListaProdutoItem `json:"produtos"`
	Mensagem string             `json:"mensagem,omitempty"`
}

// ConfirmacaoItem for confirmacao template
type ConfirmacaoItem struct {
	Nome       string  `json:"nome"`
	Quantidade string  `json:"quantidade,omitempty"`
	Preco      float64 `json:"preco"`
	Emoji      string  `json:"emoji,omitempty"`
}

// ConfirmacaoData for confirmacao template
type ConfirmacaoData struct {
	Titulo          string            `json:"titulo,omitempty"`
	Subtitulo       string            `json:"subtitulo,omitempty"`
	Itens           []ConfirmacaoItem `json:"itens"`
	ValorTotal      float64           `json:"valor_total"`
	PrevisaoEntrega string            `json:"previsao_entrega,omitempty"`
	Mensagem        string            `json:"mensagem,omitempty"`
}

// CobrancaPixData for cobranca_pix template
type CobrancaPixData struct {
	Valor        float64 `json:"valor"`
	NumeroPedido string  `json:"numero_pedido,omitempty"`
	PixPayload   string  `json:"pix_payload"`
	Expiracao    string  `json:"expiracao,omitempty"`
	Mensagem     string  `json:"mensagem,omitempty"`
}
